### ISSUE TITLE
Tensor.nbytes() returns itemsize * numel for sparse tensors

### ISSUE THREAD

**Issue by gchanan** [pytorch/pytorch#29734]
scipy sparse tensors don't have an "nbytes".  I guess you could argue for this or doing nnz * itemsize, but it seems best to just error in this case.

**Comment by vincentqb** [comment]
I'm adding the autograd label since the related pull request touches files in autograd. @gchanan please update this if there is a better label.

**Comment by vadimkantorov** [comment]
Also, instance method (or property?) Tensor.nbytes isn't searchable via docs / documented: https://pytorch.org/docs/stable/search.html?q=nbytes&check_keywords=yes&area=default

**Comment by albanD** [comment]
The problem is that if we follow numpy's definition (which we do) then the `python_tensor_megabytes` is not really correct here.
In particular, it reprensents the number of bytes these Tensors would use if they were all contiguous in memory.
Meaning that if you have only a slice of a bigger storage: you won't see that you're holding to a lot more than what you actually do.
And if you did some expand(), then you will see the full size even though you actually hold on to only a subset of that memory.

This dilemma is even bigger for sparse: what should you return? Actual memory usage? Or expected one in some hypothetical case?

So I'm not sure we want to make it public as it is not clear what it returns.
Maybe more specialized functions that report how much memory you actually use would be better?

Finally for your example, you should deduplicate Tensors that are view of each other and thus have a single shared memory buffer.

**Comment by vadimkantorov** [comment]
maybe name it more specifically to be clear if it returns only the physically viewed storage (e.g. only the slice of the storage, and not considering the expands). And have a different version returning the size of the full storage. Or maybe have a single method and have some argument options

**Comment by vadimkantorov** [comment]
> Actual memory usage?

Yes, actual memory usage - at least for non-slice tensors (i.e. `tensor.numel() == tensor._base.numel()`) where all of underlying storage is represented. Also, maybe it's worth to add a publicly exposed check function for this to complement `._is_view()` - which is True for both slice and non-slice views.

This would be useful to have "polymorphic" support for `.nbytes` for all kinds of existing tensor subclasses in core for usecases like:
- https://github.com/pytorch/pytorch/issues/91692

For sparse coo tensors the basic version could may be defined as:

`nbytes = spasecootensor._indices().numel() * spasecootensor._indices().element_size() + spasecootensor._values().numel() * spasecootensor._values().element_size()`

or simpler as:
`nbytes = spasecootensor._indices().nbytes + spasecootensor._values().nbytes`

This is also useful to have this `.nbytes` for all the quantized tensor variants etc - as a very visible demonstration of memory savings (e.g. if we compare `torch.bool` and packed `torch.bit`/`torch.bits8` if ever un/packbits gets added https://github.com/pytorch/ao/issues/292)

Currently `sparsecootensor.nbytes` returns:
`RuntimeError: nbytes is not defined for sparse tensors.  If you want the size of the constituent tensors, add the nbytes of the indices and values.  If you want the size of the  equivalent dense tensor, multiply numel() by element_size()`

IMO `nbytes` because of mention of bytes in the name assumes that we care about something close to the actual physical memory storage amount occupied by the logical elements. I realize, someone might want "logical" nbytes, but I guess currently it's mostly used for memory reporting and analysis, so I'd guess returning "physical" nbytes (or sum of nbytes for indices and values) is what's mostly reasonable for memory consumption analysis
